app/react/src/server/cra_config.js

```diff
@@ -0,0 +1,85 @@
+import semver from 'semver';
+import { normalizeCondition } from 'webpack/lib/RuleSet';
+
+export function isReactScriptsInstalled() {
+  try {
+    // eslint-disable-next-line global-require, import/no-extraneous-dependencies
+    const reactScriptsJson = require('react-scripts/package.json');
+    if (semver.lt(reactScriptsJson.version, '2.0.0')) return false;
+    return true;
+  } catch (e) {
+    return false;
+  }
+}
+
+export function getStyleRules(rules) {
+  // Extensions of style rules we're interested in
+  const extensions = ['.css', '.scss', '.sass', '.module.css', '.module.scss', '.module.sass'];
+
+  return rules.reduce((styleRules, rule) => {
+    // If at least one style extension satisfies the rule test, the rule is one
+    // we want to extract
+    if (rule.test && extensions.some(normalizeCondition(rule.test))) {
+      // If the base test is for styles, return early
+      return styleRules.concat(rule);
+    }
+
+    //  Get any style rules contained in rule.oneOf
+    if (!rule.test && rule.oneOf) {
+      styleRules.push(...getStyleRules(rule.oneOf));
+    }
+
+    // Get any style rules contained in rule.rules
+    if (!rule.test && rule.rules) {
+      styleRules.push(...getStyleRules(rule.rules));
+    }
+
+    return styleRules;
+  }, []);
+}
+
+export function getCraWebpackConfig(mode) {
+  if (mode === 'production') {
+    // eslint-disable-next-line global-require, import/no-extraneous-dependencies
+    return require('react-scripts/config/webpack.config.prod');
+  }
+
+  // eslint-disable-next-line global-require, import/no-extraneous-dependencies
+  return require('react-scripts/config/webpack.config.dev');
+}
+
+export function applyCRAWebpackConfig(baseConfig) {
+  // Remove any rules from baseConfig that test true for any one of the extensions
+  const baseRulesExcludingStyles = baseConfig.module.rules.filter(
+    rule => !rule.test || !['.css', '.scss', '.sass'].some(normalizeCondition(rule.test))
+  );
+
+  //  Load create-react-app config
+  const craWebpackConfig = getCraWebpackConfig(baseConfig.mode);
+
+  const craStyleRules = getStyleRules(craWebpackConfig.module.rules);
+
+  //  Add css minification for production
+  const plugins = [...baseConfig.plugins];
+  if (baseConfig.mode === 'production') {
+    // eslint-disable-next-line global-require, import/no-extraneous-dependencies
+    const MiniCssExtractPlugin = require('mini-css-extract-plugin');
```

Comment:
On a second thought, I think this can make problems if 'mini-css-extract-plugin' is not installed as the root package in node_modules 🤔

Reply:
Maybe we need to do something like this instead ?

```js
const miniCssExtractPlugin = craWebpackConfig.plugins.find(plugin => plugin.constructor.name === 'MiniCssExtractPlugin');
```

Reply:
I think you were right initially. Since we check to make sure `react-scripts` is installed before calling `applyCRAWebpackConfig` (I Just noticed this has all caps for CRA), then we know `mini-css-extract-plugin` is present.

Reply:
If we want to use it, we have to depend on it

Reply:
> we know mini-css-extract-plugin is present.

there's no guarantee that it's available in root `node_modules` not in `node_modules/react-scripts/node_modules`

Reply:
Looks like #4524 is broken now because of this

Reply:
#4534